chat/README.md

```diff
@@ -1,49 +1,126 @@
+
+This function is called on the local peer when a remote peer initiate a connection and starts a stream with the local peer.
+```go
+// Set a function as stream handler.
+host.SetStreamHandler("chat/1.1.0", handleStream)
+```
+
+```handleStream``` is executed for each new stream incoming to the local peer. ```s new.Stream``` is used to exchange data between local and remote peer. This example uses non blocking functions for reading anf writing from this stream.
+
+```go
+func handleStream(s net.Stream) {
+
+    // Create a buffer stream for non blocking read and write.
+    rw := bufio.NewReadWriter(bufio.NewReader(s), bufio.NewWriter(s))
+
+    go readData(rw)
+    go writeData(rw)
+
+    // stream 's' will stay open until you close it (or the other side closes it).
+}
+```
 
-On node 'B'
+3. **Initiate a new DHT Client with ```host``` as local peer.**
 
+```go
+dht := dht.NewDHTClient(ctx, host, datastore.NewMapDatastore())
 ```
-> ./chat -sp 3001
-Run ./chat -d /ip4/127.0.0.1/tcp/3001/ipfs/QmdXGaeGiVA745XorV1jr11RHxB9z4fqykm6xCUPX1aTJo
 
-2018/02/27 01:21:32 Got a new stream!
-> hi (received messages in green colour)
-> hello (sent messages in white colour)
-> no
+4. **Connect to IPFS bootstrap nodes.**
+
+These nodes are used to find nearby peers using DHT.
+
+```go
+for _, addr := range bootstrapPeers {
+
+    iaddr, _ := ipfsaddr.ParseString(addr)
+
+    peerinfo, _ := peerstore.InfoFromP2pAddr(iaddr.Multiaddr())
+
+    if err := host.Connect(ctx, *peerinfo); err != nil {
+        fmt.Println(err)
+    } else {
+        fmt.Println("Connection established with bootstrap node: ", *peerinfo)
+    }
+}
 ```
 
-On node 'A'. Replace 127.0.0.1 with <PUBLIC_IP> if node 'B' has one.
+5. **Announce your presence using a rendezvous point.**
 
+```dht.Provide``` makes this node announce that it can provide a value for the given key. Where a key in this case is ```rendezvousPoint```. Other peers will hit the same key to find other peers.
+
+```go
+if err := dht.Provide(tctx, rendezvousPoint, true); err != nil {
+    panic(err)
+}
 ```
-> ./chat -d /ip4/127.0.0.1/tcp/3001/ipfs/QmdXGaeGiVA745XorV1jr11RHxB9z4fqykm6xCUPX1aTJo
-Run ./chat -d /ip4/127.0.0.1/tcp/3001/ipfs/QmdXGaeGiVA745XorV1jr11RHxB9z4fqykm6xCUPX1aTJo
 
-This node's multiaddress:
-/ip4/0.0.0.0/tcp/0/ipfs/QmWVx9NwsgaVWMRHNCpesq1WQAw2T3JurjGDNeVNWifPS7
-> hi
-> hello
+6. **Find peers nearby.**
+
+```dht.FindProviders``` will return all the peers who have announced their presence before.
+
+```go
+peers, err := dht.FindProviders(tctx, rendezvousPoint)
 ```
 
-**NOTE: debug mode is enabled by default, debug mode will always generate same node id (on each node) on every execution. Disable debug using `--debug false` flag while running your executable.**
+7. **Connect and open streams to peers found.**
+
+Finally we connect to the peers and open stream to them.
+
+```go
+for _, p := range peers {
+    if err := host.Connect(tctx, p); err != nil {
```

Comment:
FYI, we actually stash the peer addresses in the peerstore. You can skip this step and open a stream directly (it'll automatically connect).

Reply:
👍